File: 1/2020.10.14/213.House Robber II/java/solution.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        if (nums.length == 1) {
            return nums[0];
        }
        return Math.max(rob(nums, 1, nums.length - 1), rob(nums, 0, nums.length - 2));
    }

    private int rob(int[] nums, int start, int end) {
        int first = 0, second = 0;
        for (int i = start; i <= end; i++) {
            int tmp = Math.max(second, first + nums[i]);
            first = second;
            second = tmp;
        }
        return second;
    }
}

File: 1/2020.10.14/213.House Robber II/python/solution.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        def _rob(nums):
            first = second = 0
            for num in nums:
                first, second = second, max(second, first + num)
            return second
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        return max(_rob(nums[1:]), _rob(nums[:-1]))
